### ISSUE TITLE
Rule change proposal: `no-useless-undefined` add option to skip checking function body

### ISSUE THREAD

**Issue by Nkmol** [sindresorhus/eslint-plugin-unicorn#2295]
It is quite normal for a function to return `T | undefined`, for example with the [`Map.get()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get) function. For this function it is also common to do some input checks, or return value checks, before doing something with the value. For these cases it also common to do an early exit with `return`, this is however conflicting with [eslint/consistent-return](https://eslint.org/docs/latest/rules/consistent-return). 

Currently, the rule of `no-useless-undefined` would error in the following case:
```ts
declare const map: Map<number, string>;
function valueOrUndefined(value: number) {
  return map.get(0);
}

declare const input: number | undefined;
function readInput() {
  if(input === undefined) {
    console.error("Pick a valid input value.")
    return undefined;
  }

  return valueOrUndefined(input);
}
```

I believe it is good to let the user pick is preference in this case. Personally, I like `return undefined` in his case.

----
Related to https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1199

**Comment by syyyr** [comment]
I have another example similar to this:
```typescript
type X = {
    foo: () => string | undefined
}

const x: X = {
    foo() {
        return undefined;
    },
}
```
If I keep the undefined, `no-useless-undefined` complains, if I remove it, typescript complains, that `foo` is not a void function.

**Comment by sindresorhus** [comment]
I don’t think we should add this option. The rule already ignores explicit TypeScript return types where `undefined` is part of the contract. Supporting inferred or contextual TypeScript return types would require type-aware analysis, and a broad `checkReturn` option would weaken one of the rule’s core checks. For `consistent-return`, use `treatUndefinedAsUnspecified: true` or disable that rule where TypeScript return types already enforce the behavior.
